handler.go

```diff
@@ -150,41 +133,113 @@ func (h *Handler) makeAuthenticationHandler(fn func(http.ResponseWriter, *http.R
 
 // serveQuery parses an incoming query and returns the results.
 func (h *Handler) serveQuery(w http.ResponseWriter, r *http.Request, u *User) {
-	// TODO: Authentication.
+	q := r.URL.Query()
+	p := influxql.NewParser(strings.NewReader(q.Get("q")))
 
 	// Parse query from query string.
-	urlQry := r.URL.Query()
-	_, err := influxql.NewParser(strings.NewReader(urlQry.Get("q"))).ParseQuery()
+	query, err := p.ParseQuery()
 	if err != nil {
-		h.error(w, "parse error: "+err.Error(), http.StatusBadRequest)
+		h.error(w, "error parsing query: "+err.Error(), http.StatusBadRequest)
 		return
 	}
 
-	// Retrieve database from server.
-	/*
-		db := h.server.Database(urlQry.Get(":db"))
-		if db == nil {
-			h.error(w, ErrDatabaseNotFound.Error(), http.StatusNotFound)
-			return
+	// Bootstrapping a secured cluster involves a special authentication policy.
+	// Until an admin-level user exists on the cluster one, and only one operation
+	// is permitted without authentication -- the creation of the first user.
+	// This user is also automatically granted admin-level access.
+	if h.AuthenticationEnabled && len(h.server.Users()) > 0 {
+		if len(query.Statements) != 1 {
+			_, ok := query.Statements[0].(*influxql.CreateUserStatement)
+			if !ok {
+				h.error(w, "initial admin user does not exist", http.StatusUnauthorized)
+				return
+			}
 		}
-	*/
+	}
 
-	// Parse the time precision from the query params.
-	/*
-		precision, err := parseTimePrecision(urlQry.Get("time_precision"))
-		if err != nil {
-			h.error(w, err.Error(), http.StatusBadRequest)
-			return
-		}
-	*/
+	for _, s := range query.Statements {
+		switch c := s.(type) {
+		case *influxql.CreateDatabaseStatement:
+			err = h.server.CreateDatabase(c.Name)
+			if err != nil {
+				if err == ErrDatabaseExists {
+					h.error(w, "database exists", http.StatusConflict)
+				} else {
+					h.error(w, "error executing query: "+err.Error(), http.StatusInternalServerError)
+				}
+			} else {
+				w.WriteHeader(http.StatusCreated)
+			}
```

Comment:
Can you consolidate the error checks and also remove indentation from the regular non-error flow:

```go
if err := h.server.CreateDatabase(c.Name); err == ErrDatabaseExists {
	h.error(w, err.Error(), http.StatusConflict)
	continue
} else if err != nil {
	h.error(w, "error executing query: "+err.Error(), http.StatusInternalServerError)
	continue
}
w.WriteHeader(http.StatusCreated)
```